receiver/prometheusreceiver/internal/metrics_adjuster.go

```diff
@@ -0,0 +1,220 @@
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"go.uber.org/zap"
+	"strings"
+)
+
+type timeseriesinfo struct {
+	initial  *metricspb.TimeSeries
+	previous *metricspb.TimeSeries
+}
+
+type metricsInstanceMap map[string]*timeseriesinfo
+
+func newMetricsInstanceMap() *metricsInstanceMap {
+	mim := metricsInstanceMap(make(map[string]*timeseriesinfo))
+	return &mim
+}
+
+func (mim *metricsInstanceMap) get(metric *metricspb.Metric, values []*metricspb.LabelValue) *timeseriesinfo {
+	name := metric.GetMetricDescriptor().GetName()
+	sig := getSignature(name, values)
+	tsi, ok := (*mim)[sig]
+	if !ok {
+		tsi = &timeseriesinfo{}
+		(*mim)[sig] = tsi
+	}
+	return tsi
+}
+
+// create a unique signature consisting of a metric's name and label values
+func getSignature(name string, values []*metricspb.LabelValue) string {
+	labelValues := make([]string, 0, len(values))
+	for _, label := range values {
+		if label.GetValue() != "" {
+			labelValues = append(labelValues, label.GetValue())
+		}
+	}
+	return fmt.Sprintf("%s,%s", name, strings.Join(labelValues, ","))
+}
+
+// JobsMap maps from a job instance to a map of metric instances for the job.
+type JobsMap map[string]*metricsInstanceMap
+
+// NewJobsMap creates a new (empty) JobsMap.
+func NewJobsMap() *JobsMap {
+	jm := JobsMap(make(map[string]*metricsInstanceMap))
+	return &jm
+}
+
+func (jm *JobsMap) get(job, instance string) *metricsInstanceMap {
+	sig := job + ":" + instance
+	metricsMap, ok := (*jm)[sig]
+	if !ok {
+		metricsMap = newMetricsInstanceMap()
+		(*jm)[sig] = metricsMap
+	}
+	return metricsMap
+}
+
+// MetricsAdjuster takes a map from a metric instance to the initial point in the metrics instance
+// and provides AdjustMetrics, which takes a sequence of metrics and adjust their values based on
+// the initial points.
+type MetricsAdjuster struct {
+	metricsMap *metricsInstanceMap
+	logger     *zap.SugaredLogger
+}
+
+// NewMetricsAdjuster is a constructor for MetricsAdjuster.
+func NewMetricsAdjuster(metricsMap *metricsInstanceMap, logger *zap.SugaredLogger) *MetricsAdjuster {
+	return &MetricsAdjuster{
+		metricsMap: metricsMap,
+		logger:     logger,
+	}
+}
+
+// AdjustMetrics takes a sequence of metrics and adjust their values based on the initial points in the
+// metricsMap. If the metric is the first point in the timeseries, or the timeseries has been reset, it is
+// removed from the sequence and added to the the metricsMap.
+func (ma *MetricsAdjuster) AdjustMetrics(metrics []*metricspb.Metric) []*metricspb.Metric {
+	var adjusted = make([]*metricspb.Metric, 0, len(metrics))
+	for _, metric := range metrics {
+		if ma.adjustMetric(metric) {
+			adjusted = append(adjusted, metric)
+		}
+	}
+	return adjusted
+}
+
+// returns true if at least one of the metric's timeseries was adjusted and false if all of the timeseries are an initial occurence or a reset.
+// Types of metrics returned supported by prometheus:
```

Comment:
How about gauge and cumulative int64 values?

Reply:
These kinds of metrics aren't generated by the Prometheus -> OC Metrics translation so I didn't add support for them.